x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/settings/TransportUpdateSecuritySettingsAction.java

```diff
@@ -0,0 +1,167 @@
+
+    private final MetadataUpdateSettingsService updateSettingsService;
+
+    @Inject
+    public TransportUpdateSecuritySettingsAction(
+        TransportService transportService,
+        ClusterService clusterService,
+        ThreadPool threadPool,
+        ActionFilters actionFilters,
+        MetadataUpdateSettingsService metadataUpdateSettingsService,
+        IndexNameExpressionResolver indexNameExpressionResolver
+    ) {
+        super(
+            UpdateSecuritySettingsAction.NAME,
+            transportService,
+            clusterService,
+            threadPool,
+            actionFilters,
+            UpdateSecuritySettingsAction.Request::new,
+            indexNameExpressionResolver,
+            AcknowledgedResponse::readFrom,
+            ThreadPool.Names.SAME
+        );
+        this.updateSettingsService = metadataUpdateSettingsService;
+    }
+
+    @Override
+    protected void masterOperation(
+        Task task,
+        UpdateSecuritySettingsAction.Request request,
+        ClusterState state,
+        ActionListener<AcknowledgedResponse> listener
+    ) {
+
+        List<UpdateSettingsClusterStateUpdateRequest> settingsUpdateRequests = Stream.of(
+            createUpdateSettingsRequest(
+                SECURITY_MAIN_ALIAS,
+                Settings.builder().loadFromMap(request.mainIndexSettings()).build(),
+                request.timeout(),
+                request.masterNodeTimeout(),
+                state
+            ),
+            createUpdateSettingsRequest(
+                SECURITY_TOKENS_ALIAS,
+                Settings.builder().loadFromMap(request.tokensIndexSettings()).build(),
+                request.timeout(),
+                request.masterNodeTimeout(),
+                state
+            ),
+            createUpdateSettingsRequest(
+                SECURITY_PROFILE_ALIAS,
+                Settings.builder().loadFromMap(request.profilesIndexSettings()).build(),
+                request.timeout(),
+                request.masterNodeTimeout(),
+                state
+            )
+        ).filter(Optional::isPresent).map(Optional::get).toList();
+        if (settingsUpdateRequests.isEmpty() == false) {
+            ActionListener<AcknowledgedResponse> groupedListener = new GroupedActionListener<>(
+                settingsUpdateRequests.size(),
+                ActionListener.wrap((responses) -> {
+                    listener.onResponse(AcknowledgedResponse.of(responses.stream().allMatch(AcknowledgedResponse::isAcknowledged)));
+                }, listener::onFailure)
+            );
+            settingsUpdateRequests.forEach(req -> updateSettingsService.updateSettings(req, groupedListener));
+        } else {
+            // All settings blocks were empty, which doesn't do anything, so this was probably a mistake
+            throw new IllegalArgumentException("No settings to update");
```

Comment:
Good call, I've added an assertion here to clarify that this ought to be impossible. I've left the `throw` in place, since we still need that for control flow purposes if assertions are disabled and the "impossible" happens.